src/niri.rs

```diff
@@ -891,20 +911,26 @@ impl State {
             );
 
             // Tell the windows their new focus state for window rule purposes.
-            if let KeyboardFocus::Layout {
+            let previous_focus_id = if let KeyboardFocus::Layout {
                 surface: Some(surface),
             } = &self.niri.keyboard_focus
             {
                 if let Some((mapped, _)) = self.niri.layout.find_window_and_output_mut(surface) {
                     mapped.set_is_focused(false);
+                    Some(LayoutElement::id(mapped).clone())
+                } else {
+                    None
                 }
-            }
+            } else {
+                None
+            };
             if let KeyboardFocus::Layout {
                 surface: Some(surface),
             } = &focus
             {
                 if let Some((mapped, _)) = self.niri.layout.find_window_and_output_mut(surface) {
                     mapped.set_is_focused(true);
+                    mapped.set_previous_focus_id(previous_focus_id)
```

Comment:
Doing it here, rather than in the layout, definitely sounds much simpler, but it means that in cases like "open keyboard-grabbing layer-shell, switch to a different layout window, close keyboard-grabbing layer-shell" the previous window will be lost. But I guess it's not really a big deal?

Reply:
Hm, because `previous_focus_id` actually forms a graph of sorts (each window pointing to another window), maybe we can just follow the edges until we find a `None` or a window that still exists?

That would solve the question of "what to do when the previous window has closed", and also handles this case gracefully.

Is there an easy way to check if a `Window` still exists on any workspace?

Reply:
Well, you won't be able to follow previous focus id past a None or a window that no longer exists (since when a window closes, you lose the previous id stored on it). I don't think that is a big problem though.

Maybe for now it's sufficient to just skip setting the previous focus id if it is None? Then if you open and close layer-shell, the window won't lose its previous focus id. If you switch to a different window while focused on a layer-shell, you'll end up with an unexpected previous focus id, but I guess that doesn't happen very often?

Reply:
> since when a window closes, you lose the previous id stored on it

Oh, right, window != mapped. I had that wrong in my mind.

I'm fine with your suggestion, this probably shouldn't cause too much unexpected behaviour. And it could be improved later on :shrug: